package/MDAnalysis/analysis/MSD.py

```diff
@@ -0,0 +1,452 @@
+                  "from choice of start_frame/stop_frame or insufficient RAM"
+                  "when loading the trajectory.")
+
+    def _select(self):
+        """Generates list of atom groups that satisfy select_list
+           strings
+
+        Returns
+        -------
+        selections : list
+            list of atom groups
+        """
+        selections = []
+        for i, sel in enumerate(self.select_list):
+            selections.append(self.universe.select_atoms(sel))
+            if selections[i].n_atoms == 0:
+                selections.pop(i)
+                selections.append(None)
+        
+        return selections
+
+    def _init_pos_deque(self):
+        """Initializes lists necessary for MSD calculations at the
+           first restart.
+
+        Deques with maxlen=len_msd are used to conserve memory.
+
+        Returns:
+        --------
+        pos : list
+            contains position information, organized:
+            selection X frame X atom X (x, y, z)
+        set_list : list
+            contains set of atoms which satisfy each selection at each
+            frame, organized:
+            selection X frame X set of atom ids
+        dict_list : list
+            contains dictionaries which map atom ids to corresponding
+            pos index, organized:
+            selection X frame X dict
+        """
+
+        n_sel = len(self.select_list)
+        
+        # dictionaries with key: (id) and value: (array index)
+        # selection X frame X dictionary
+        dict_list = [[dict() for j in range(self.len_msd)]
+                     for i in range(n_sel)]
+        # atom ids which satisfy the selection at each frame
+        set_list = [[set() for j in range(self.len_msd)]
+                    for i in range(n_sel)]
+
+        # pre-allocate position array
+        pos = [[np.array([]) for j in range(self.len_msd)]
+               for i in range(n_sel)]
+
+        print("Populating deque...")
+        for ts in self.universe.trajectory[self.start_frame:self.start_frame
+                                           + self.len_msd]:
+            this_frame = ts.frame
+            selections = self._select()
+
+            for i, sel in enumerate(selections):
+                if sel is not None:  # if there is data
+                    pos[i][this_frame-self.start_frame] = sel.positions.copy()
+                    
+                    temp_list = sel.atoms.ids
+                    # store set of atom ids which satisfy selection
+                    set_list[i][this_frame-self.start_frame] = set(temp_list)
+                    # link atom id to array index
+                    for j in range(len(temp_list)):
+                        dict_list[i][this_frame-self.start_frame][temp_list[j]] = j
+                    
+            if this_frame % (self.len_msd/10) == 0:
+                print("Frame: "+str(this_frame))
+        # converting lists to deques
```

Comment:
This seems very fragile. You would need to ensure that you to exactly `len_msd` appends in `update_deque` to remove any stale data points in them. If you do to many or to little the MSD will be slightly off.  It would be more robust if you leave this a list and empty them in `update_deque` and check there that they have been filled correctly.